File: CSharpCodeBase/components/services/spells.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class spells {
 var SpellSystem = Class({init=function(self)
    this.currentSpellAction = null;
    this.statuses = {}
 }})
 public void SpellSystem:GetSpell(spellName){
   var spell = GameController.database:Get("spells", spellName);
   spell.name = spellName  ;
   return spell;
 }
 public void SpellSystem:StartSpell(spellName){
   var spell = self:GetSpell(spellName);
   var spellActionClass = require(spell.class);
   this.currentSpellAction = spellActionClass(spellName);
   this.currentSpellAction:BeginDraw();
 }
 public void SpellSystem:DropSpell(){
   if(this.currentSpellAction != null  ){ 
     this.currentSpellAction:StopDraw();
   }
 }
 public void SpellSystem:ApplySpell(spellName, position)   {
   if(spellName == ""  &&  spellName == null  ){  return false }
   var spell = self:GetSpell(spellName);
   var spellActionClass = require(spell.class);
   if(spellActionClass.CanApply(position)  ){ 
     //if this.currentSpellAction == null  ){ 
       this.currentSpellAction = spellActionClass(spellName)      ;
     //end
     if(position  ){                
       this.currentSpellAction.position = position;
     }else{      
       this.currentSpellAction.position = null;
     }
     GameController.player.playerController:GetCurrentSlot():OnSpellCast(this.currentSpellAction)        ;
     return true;
   }else{    
     return false;
   }
 }
 public void SpellSystem:SaveStatuses(){
   var currentHero = GameController.player.playerController:GetCurrentSlot();
   if(not this.statuses[currentHero.name]  ){ 
     this.statuses[currentHero.name] = {}    
   }
   for(i=1, 4 ){        
     var status = GameController.ui.mainInterface:GetSpellStatus(i);
     this.statuses[currentHero.name][i] = status;
     if(status.mode == "active"  ||  status.mode == "recharge"  ){       
       currentHero:SetTimer(i, status.timeout);
     }
   }
 }
 public void SpellSystem:LoadStatuses(){
   var currentHero = GameController.player.playerController:GetCurrentSlot();
   if(this.statuses[currentHero.name]  ){ 
     for(i=1, 4 ){
       GameController.ui.mainInterface:SetSpellStatus(i, this.statuses[currentHero.name][i]);
       this.statuses[currentHero.name][i] = {}
     }
     currentHero.timeManager:Clear();
     this.statuses[currentHero.name] = null;
   }  
 }
 public void SpellSystem:TimerUpdate(hero, action, timeout)  {
   if(this.statuses[hero]  ){ 
     this.statuses[hero][action]["timeout"] = timeout;
   }
 }
 public void SpellSystem:SetupSpells(spellNames)  {
   GameController.ui.mainInterface:ResetSpells()  ;
   if(spellNames != null  ){     
     for(k, v in pairs(spellNames) ){
       var spell = GameController.database:Get("spells", v);
       GameController.ui.mainInterface:SetupSpell(spell.action, v, spell.icon, spell.recharge_timeout  ||  1, spell.useCursor, spell.action_timeout)      ;
     }
   }
   self:LoadStatuses();
 }
 ystem}}

File: CSharpCodeBase/components/services/spells.lua
local Class = require 'utils.hump.class'

local SpellSystem = Class({init=function(self)
   self.currentSpellAction = nil
   self.statuses = {}
end})

function SpellSystem:GetSpell(spellName)
  local spell = GameController.database:Get('spells', spellName)
  spell.name = spellName  
  return spell
end

function SpellSystem:StartSpell(spellName)
  local spell = self:GetSpell(spellName)
  local spellActionClass = require(spell.class)
  self.currentSpellAction = spellActionClass(spellName)
  self.currentSpellAction:BeginDraw()
end

function SpellSystem:DropSpell()
  if self.currentSpellAction ~= nil then
    self.currentSpellAction:StopDraw()
  end
end

function SpellSystem:ApplySpell(spellName, position)   
  if spellName == '' and spellName == nil then return false end
  local spell = self:GetSpell(spellName)
  local spellActionClass = require(spell.class)
  if spellActionClass.CanApply(position) then
    --if self.currentSpellAction == nil then
      self.currentSpellAction = spellActionClass(spellName)      
    --end
    if position then               
      self.currentSpellAction.position = position
    else      
      self.currentSpellAction.position = nil
    end
    GameController.player.playerController:GetCurrentSlot():OnSpellCast(self.currentSpellAction)        
    return true
  else    
    return false
  end
end

function SpellSystem:SaveStatuses()
  local currentHero = GameController.player.playerController:GetCurrentSlot()
  if not self.statuses[currentHero.name] then
    self.statuses[currentHero.name] = {}    
  end
  for i=1, 4 do        
    local status = GameController.ui.mainInterface:GetSpellStatus(i)
    self.statuses[currentHero.name][i] = status
    if status.mode == "active" or status.mode == "recharge" then      
      currentHero:SetTimer(i, status.timeout)
    end
  end
end

function SpellSystem:LoadStatuses()
  local currentHero = GameController.player.playerController:GetCurrentSlot()
  if self.statuses[currentHero.name] then
    for i=1, 4 do
      GameController.ui.mainInterface:SetSpellStatus(i, self.statuses[currentHero.name][i])
      self.statuses[currentHero.name][i] = {}
    end
    currentHero.timeManager:Clear()
    self.statuses[currentHero.name] = nil
  end  
end

function SpellSystem:TimerUpdate(hero, action, timeout)  
  if self.statuses[hero] then
    self.statuses[hero][action]['timeout'] = timeout
  end
end

function SpellSystem:SetupSpells(spellNames)  
  GameController.ui.mainInterface:ResetSpells()  
  if spellNames ~= nil then    
    for k, v in pairs(spellNames) do
      local spell = GameController.database:Get('spells', v)
      GameController.ui.mainInterface:SetupSpell(spell.action, v, spell.icon, spell.recharge_timeout or 1, spell.useCursor, spell.action_timeout)      
    end
  end
  self:LoadStatuses()
end

return SpellSystem